File: go/pkg/models/quest_archetype.go
package models

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

type QuestArchetype struct {
	ID                            uuid.UUID                   `json:"id"`
	Name                          string                      `json:"name"`
	Description                   string                      `json:"description"`
	Category                      string                      `json:"category" gorm:"column:category;default:'side'"`
	QuestGiverCharacterID         *uuid.UUID                  `json:"questGiverCharacterId,omitempty" gorm:"column:quest_giver_character_id;type:uuid"`
	QuestGiverCharacter           *Character                  `json:"questGiverCharacter,omitempty" gorm:"foreignKey:QuestGiverCharacterID"`
	RequiredStoryFlags            StringArray                 `json:"requiredStoryFlags" gorm:"column:required_story_flags;type:jsonb;default:'[]'"`
	SetStoryFlags                 StringArray                 `json:"setStoryFlags" gorm:"column:set_story_flags;type:jsonb;default:'[]'"`
	ClearStoryFlags               StringArray                 `json:"clearStoryFlags" gorm:"column:clear_story_flags;type:jsonb;default:'[]'"`
	QuestGiverRelationshipEffects CharacterRelationshipState  `json:"questGiverRelationshipEffects" gorm:"column:quest_giver_relationship_effects;type:jsonb;default:'{}'"`
	AcceptanceDialogue            StringArray                 `json:"acceptanceDialogue,omitempty" gorm:"type:jsonb"`
	ImageURL                      string                      `json:"imageUrl"`
	DifficultyMode                QuestDifficultyMode         `json:"difficultyMode" gorm:"column:difficulty_mode"`
	Difficulty                    int                         `json:"difficulty" gorm:"default:1"`
	MonsterEncounterTargetLevel   int                         `json:"monsterEncounterTargetLevel" gorm:"column:monster_encounter_target_level;default:1"`
	DefaultGold                   int                         `json:"defaultGold"`
	RewardMode                    RewardMode                  `json:"rewardMode" gorm:"column:reward_mode"`
	RandomRewardSize              RandomRewardSize            `json:"randomRewardSize" gorm:"column:random_reward_size"`
	RewardExperience              int                         `json:"rewardExperience" gorm:"column:reward_experience"`
	RecurrenceFrequency           *string                     `json:"recurrenceFrequency,omitempty"`
	MaterialRewards               BaseMaterialRewards         `json:"materialRewards" gorm:"column:material_rewards_json;type:jsonb;default:'[]'"`
	CharacterTags                 StringArray                 `json:"characterTags" gorm:"column:character_tags;type:jsonb"`
	InternalTags                  StringArray                 `json:"internalTags" gorm:"column:internal_tags;type:jsonb"`
	CreatedAt                     time.Time                   `json:"createdAt"`
	UpdatedAt                     time.Time                   `json:"updatedAt"`
	DeletedAt                     gorm.DeletedAt              `json:"deletedAt"`
	Root                          QuestArchetypeNode          `json:"root"`
	RootID                        uuid.UUID                   `json:"rootId"`
	ItemRewards                   []QuestArchetypeItemReward  `json:"itemRewards" gorm:"foreignKey:QuestArchetypeID"`
	SpellRewards                  []QuestArchetypeSpellReward `json:"spellRewards" gorm:"foreignKey:QuestArchetypeID"`
}

File: js/packages/types/src/questArchetype.ts
import { InventoryItem } from './inventoryItem';
import { LocationArchetype } from './locationArchetype';
import { QuestDifficultyMode } from './questDifficulty';
import { CharacterRelationshipState, QuestMaterialReward } from './quest';
import { Spell } from './spell';
import { Character } from './character';
import type { DialogueMessage } from './characterAction';
import type {
  ExpositionMaterialReward,
  ExpositionRewardMode,
  ExpositionRandomRewardSize,
} from './exposition';

export interface QuestArchetypeChallenge {
  id: string;
  createdAt: Date;
  updatedAt: Date;
  deletedAt?: Date;
  challengeTemplateId?: string | null;
  challengeTemplate?: QuestArchetypeChallengeTemplate | null;
  proficiency?: string | null;
  difficulty?: number | null;
  unlockedNodeId?: string;
  unlockedNode?: QuestArchetypeNode;
}

export interface QuestArchetypeChallengeTemplate {
  id: string;
  locationArchetypeId: string;
  question: string;
  description?: string;
  submissionType?: string;
  difficulty?: number | null;
  proficiency?: string | null;
}

export type QuestArchetypeNodeType =
  | 'location'
  | 'monster_encounter'
  | 'scenario'
  | 'exposition';

export interface QuestArchetypeNodeExpositionItemReward {
  inventoryItemId: number;
  quantity: number;
}

export interface QuestArchetypeNodeExpositionSpellReward {
  spellId: string;
}

export interface QuestArchetypeItemReward {
  id?: string;
  questArchetypeId?: string;
  inventoryItemId: number;
  inventoryItem?: InventoryItem;
  quantity: number;
}

export interface QuestArchetypeSpellReward {
  id?: string;
  questArchetypeId?: string;
  spellId: string;
  spell?: Spell;
}

export interface QuestArchetypeNode {
  id: string;
  createdAt: Date;
  updatedAt: Date;
  deletedAt?: Date;
  nodeType?: QuestArchetypeNodeType;
  locationArchetype?: LocationArchetype | null;
  locationArchetypeId?: string | null;
  scenarioTemplateId?: string | null;
  monsterTemplateIds?: string[];
  targetLevel?: number | null;
  encounterProximityMeters?: number | null;
  expositionTitle?: string | null;
  expositionDescription?: string | null;
  expositionDialogue?: DialogueMessage[];
  expositionRewardMode?: ExpositionRewardMode;
  expositionRandomRewardSize?: ExpositionRandomRewardSize;
  expositionRewardExperience?: number | null;
  expositionRewardGold?: number | null;
  expositionMaterialRewards?: ExpositionMaterialReward[];
  expositionItemRewards?: QuestArchetypeNodeExpositionItemReward[];
  expositionSpellRewards?: QuestArchetypeNodeExpositionSpellReward[];
  challenges: QuestArchetypeChallenge[];
  difficulty?: number | null;
}

export interface QuestArchetype {
  id: string;
  name: string;
  description: string;
  category?: 'side' | 'main_story';
  questGiverCharacterId?: string | null;
  questGiverCharacter?: Character | null;
  acceptanceDialogue?: DialogueMessage[];
  imageUrl?: string;
  difficultyMode?: QuestDifficultyMode;
  difficulty?: number;
  monsterEncounterTargetLevel?: number;
  defaultGold: number;
  rewardMode?: 'explicit' | 'random';
  randomRewardSize?: 'small' | 'medium' | 'large';
  rewardExperience?: number;
  recurrenceFrequency?: string | null;
  materialRewards?: QuestMaterialReward[];
  requiredStoryFlags?: string[];
  setStoryFlags?: string[];
  clearStoryFlags?: string[];
  questGiverRelationshipEffects?: CharacterRelationshipState;
  characterTags?: string[];
  internalTags?: string[];
  createdAt: Date;
  updatedAt: Date;
  deletedAt?: Date;
  root: QuestArchetypeNode;
  rootId: string;
  itemRewards?: QuestArchetypeItemReward[];
  spellRewards?: QuestArchetypeSpellReward[];
}
